File: hw1/main.py
'''
Author: Cody Costa
Date:   7/14/2025

'''

import cv2
import matplotlib.pyplot as plt
import numpy as np


''' PROBLEM 1:  WARMUP, DISPLAY GS AND COLOR IMG '''
# filename to open
LENA = 'lena.webp'

# read image to memory
img = cv2.imread(LENA)

# display image
cv2.imshow('image', img)

# convert to grayscale
img_gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)

# display gray image
cv2.imshow('gray image', img_gray)

# await key press to close program
cv2.waitKey(0)
cv2.destroyAllWindows()



''' PROBLEM 2:  SOBEL AND PREWITT EDGE DETECTION '''
# Sobel edge detection using OpenCV built in functions

# apply 3x3 x and y Sobel kernels
sobel_x = cv2.Sobel(img_gray, cv2.CV_64F, 1, 0, ksize=3)
sobel_y = cv2.Sobel(img_gray, cv2.CV_64F, 0, 1, ksize=3)

# calculate magnitude and normalize results into an 8 bit image
sobel_mag = np.hypot(sobel_x, sobel_y)
sobel_mag = cv2.normalize(sobel_mag, None, 0, 255, cv2.NORM_MINMAX).astype(np.uint8)



# Prewitt edge detection using manual kernel and filter2D

# apply x and y Prewitt kernels 
prewitt_x = np.array([[-1, 0, 1],
                      [-1, 0, 1],
                      [-1, 0, 1]], dtype=np.float32)
# prewitt_y = np.array([[1, 1, 1],
#                       [0, 0, 0],
#                       [-1, -1, -1]], dtype=np.float32)
prewitt_y = prewitt_x.T

# apply convolution
px = cv2.filter2D(img_gray.astype(np.float32), -1, prewitt_x)
py = cv2.filter2D(img_gray.astype(np.float32), -1, prewitt_y)

# calculate magnitude and normalize results into an 8 bit image
prewitt_mag = np.hypot(px, py)
prewitt_mag = cv2.normalize(prewitt_mag, None, 0, 255, cv2.NORM_MINMAX).astype(np.uint8)



# display results
combined = np.hstack((sobel_mag, prewitt_mag))

plt.figure(figsize=(10, 5))
plt.imshow(combined, cmap='gray')
plt.title('Sobel (left) vs Prewitt (right)')
plt.axis('off')
plt.show()


import enhancement_pipline

''' PROBLEM 3:  IMAGE ENHANCEMENT PIPELINE '''
# new photo to load
BONES = 'Fig3-46a.tif'

# accessory file created to handle enhancement logic pipeline
enhancement_pipline.enhance_img(BONES)


''' PROBLEM 4:  IMAGE ENHANCEMENT PIPELINE '''
# new photo to load
MRI = 'Image9.jpg'

# accessory file created to handle enhancement logic pipeline
enhancement_pipline.enhance_img(MRI)

File: hw1/main.m
% Author:   Cody Costa
% Date:     7/15/2025


%% Images to use for assignment

LENA = 'lena.jpg';
BONES = 'Fig3-46a.tif';
MRI = 'Image9.jpg';


%% PROBLEM 1: Load and display GS and Color img
img = imread(LENA);
figure;
imshow(img)

gray = rgb2gray(img);
figure;
imshow(gray)


%% PROBLEM 2: Sobel and Prewitt edge detection
gray = rgb2gray(imread(LENA));

% Sobel
sx = [-1 0 1; 
      -2 0 2; 
      -1 0 1];
sy = sx';

Gx = conv2(double(gray), sx, 'same');
Gy = conv2(double(gray), sy, 'same');

G_mag = sqrt(Gx.^2 + Gy.^2);
edge_sobel = uint8(255 * mat2gray(G_mag));

figure;
imshow(edge_sobel)
title('Sobel Edge');

% Prewitt
px = [-1 0 1;
      -1 0 1;
      -1 0 1];
py = px';

Px = conv2(double(gray), px, 'same');
Py = conv2(double(gray), py, 'same');

P_mag = sqrt(Px.^2 + Py.^2);
edge_prewitt = uint8(255 * mat2gray(P_mag));

figure;
imshow(edge_prewitt)
title('Prewitt Edge');


%% PROBLEM 3: Image Process Pipeline
gray = imread(BONES);
G = im2double(gray);
figure;
imshow(G)
title('Base Gray Img')

% take laplacian of image
laplacianKernel = [0 1 0; 
                   1 -4 1; 
                   0 1 0];
Laplacian = imfilter(G, laplacianKernel, "replicate");
figure;
imshow(Laplacian, [])
title('Laplacian')

% sharpen image by subtracting laplacian from original
sharpened = G - Laplacian;
figure;
imshow(sharpened, [])
title('Sharpened')

% apply Sobel edge to original image
sobelBinary = edge(G, "sobel");
figure;
imshow(sobelBinary, [])
title('Sobel Edge')

% apply 5x5 blur to sobel edge image
avgKernel = fspecial('average', [5 5]);
smoothedBinary = imfilter(double(sobelBinary), avgKernel, 'replicate');
figure;
imshow(smoothedBinary, [])
title('Smoothed')

% multiply sharpened with blurred
Mask = sharpened .* smoothedBinary;
figure;
imshow(Mask, [])
title('Product')

% sum of original img with Mask
SharpMask = G + Mask;
figure;
imshow(SharpMask, [])
title('Sharp Mask')

% power transform to Sharpened Mask photo
gamma = 0.4;
c = 1.0;

IMG_GAMMA = c * (SharpMask .^ gamma);
figure;
imshow(IMG_GAMMA, [])
title('Gamma Transform')


%% PROBLEM 4: Image Process Pipeline
gray = imread(MRI);
G = im2double(gray);
figure;
imshow(G)
title('Base Gray Img')

% take laplacian of image
laplacianKernel = [0 1 0; 
                   1 -4 1; 
                   0 1 0];
Laplacian = imfilter(G, laplacianKernel, "replicate");
figure;
imshow(Laplacian, [])
title('Laplacian')

% sharpen image by subtracting laplacian from original
sharpened = G - Laplacian;
figure;
imshow(sharpened, [])
title('Sharpened')

% apply Sobel edge to original image
sobelBinary = edge(G, "sobel");
figure;
imshow(sobelBinary, [])
title('Sobel Edge')

% apply 5x5 blur to sobel edge image
avgKernel = fspecial('average', [5 5]);
smoothedBinary = imfilter(double(sobelBinary), avgKernel, 'replicate');
figure;
imshow(smoothedBinary, [])
title('Smoothed')

% multiply sharpened with blurred
Mask = sharpened .* smoothedBinary;
figure;
imshow(Mask, [])
title('Product')

% sum of original img with Mask
SharpMask = G + Mask;
figure;
imshow(SharpMask, [])
title('Sharp Mask')

% power transform to Sharpened Mask photo
gamma = 0.4;
c = 1.0;

IMG_GAMMA = c * (SharpMask .^ gamma);
figure;
imshow(IMG_GAMMA, [])
title('Gamma Transform')


%% close all figures
close all